mapper/collectd_mapper/src/batcher.rs

```diff
@@ -1,171 +1,199 @@
-                    let message_batch_result = self
-                        .build_message_batch_with_timeout(message, timestamp, messages.as_mut())
-                        .await;
-
-                    match message_batch_result {
-                        Ok(message_batch) => {
-                            // Send the current batch to the batch processor
-                            let _ = self.sender.send(message_batch).map_err(|err| {
-                                error!("Error while publishing a message batch: {}", err)
-                            });
-                        }
-                        Err(err) => {
-                            error!("Error while building a message batch: {}", err);
-                        }
-                    }
-                }
-                None => {
-                    //If the message batching loop returns, it means the MQTT connection has closed
-                    error!("MQTT connection closed. Retrying...");
-                }
+            let now = self.clock.now();
+            let retire_groups_action = self.batcher.retire_groups(now);
+
+            for retired_group in retire_groups_action.retired_groups.iter() {
+                // Send the current batch to the batch processor
+                let _ = group_messages(retired_group)
+                    .map_err(|err| error!("Error while grouping the message batch: {}", err))
+                    .and_then(|group| {
+                        self.sender.send(group).map_err(|err| {
+                            error!("Error while publishing a message batch: {}", err)
+                        })
+                    });
             }
-        }
-    }
 
-    async fn build_message_batch_with_timeout(
-        &self,
-        first_message: Message,
-        first_message_timestamp: Timestamp,
-        messages: &mut dyn MqttMessageStream,
-    ) -> Result<MeasurementGrouper, DeviceMonitorError> {
-        let collectd_message = CollectdMessage::parse_from(&first_message)?;
-        let mut message_batch =
-            MessageBatch::start_batch(collectd_message, first_message_timestamp)?;
+            let next_notification_at = retire_groups_action
+                .next_check_at
+                .unwrap_or(now + chrono::Duration::hours(24));
 
-        // Creates a sleep timer future handler and does not await here
-        // for sleep to finish, but inside the select loop
-        let sleep = time::sleep(self.batching_window);
-        tokio::pin!(sleep);
+            // To avoid negative durations (which is not supported by std::time::Duration), fall back
+            // to use a small timeout of 20 ms. Also use at least 20ms to avoid very small timeouts
+            // which might in the worst case end up in a busy loop.
+            let delta = next_notification_at - now;
+            let t20ms = std::time::Duration::from_millis(20);
```

Comment:
This timeout will have to be derived based on `maximum acceptable delay` instead of a hardcoded value.